### PR TITLE
Badge: Add new 'type' variations and best practices documentation

### PR DESCRIPTION
**Pull request by ayeshakmaz**
#### What changed?

Update Badge to accept various `type` options for new styling options

All the color changes are part of the Badge changes, since we needed to update the colors in dark mode and update the text colors for contrast on badges

Screen Shot 2022-03-09 at 12 15 57 AM
Screen Shot 2022-03-09 at 12 16 03 AM

#### Why?

To allow Badge to indicate a wider range of information, we've added various styling options through the `type` prop, and added best practices around the new options

### Links

- [Jira](https://jira.pinadmin.com/browse/GESTALT-3620)
- [Figma](https://www.figma.com/file/XLZQMmX6NKdLOwA0inZY5L/Component-Discovery?node-id=220%3A1052)

### Checklist

- [x] Added unit and Flow Tests
- [x] Added documentation + accessibility tests
- [x] Verified accessibility: keyboard & screen reader interaction
- [x] Checked dark mode, responsiveness, and right-to-left support
- [x] Checked stakeholder feedback (e.g. Gestalt designers)

### CODE REVIEW

File: packages/gestalt/src/Badge.js [hunk 1]
```diff
@@ -16,20 +15,33 @@ type Props = {|
...
-export default function Badge({ position = 'middle', text }: Props): Node {
-  const { name: colorSchemeName } = useColorScheme();
+export default function Badge({ position = 'middle', text, type = 'info' }: Props): Node {
+  const TYPE_COLOR_MAP = {
+    'info': 'infoBase',
+    'error': 'errorBase',
+    'warning': 'warningBase',
+    'success': 'successBase',
+    'neutral': 'tertiary',
+    'darkWash': 'washDark',
```
Comment (AlbertCarreras) [inline packages/gestalt/src/Badge.js:38]: interesting that they are reversed haha I wonder if this could be prevented

**Comment by AlbertCarreras** [comment]
nit: I'd have broken split this PR in 3: 

- The docs lowercase changes (they add too many pages to review)
- The bagde changes
- The color value changes

I feel the PR title don't describe all these different changes,

**Comment by ayeshakmaz** [comment]
@AlbertCarreras All the color changes are part of the Badge changes, since we needed to update the colors in dark mode and update the text colors for contrast on badges. And yeah, sorry about the extra pages, just wanted to fix those asap since we've fixed that issue everywhere else already

File: packages/gestalt/src/Badge.js [hunk 2]
```diff
@@ -16,20 +15,33 @@ type Props = {|
    * Text displayed inside of the Badge. Sentence case is preferred.
    */
   text: string,
+  /**
+   * Determines the style of the badge.
```
Comment (AlbertCarreras) [inline packages/gestalt/src/Badge.js:19]: I'd add See [type](..URL.../badge#Type) variant to learn more.

File: docs/pages/badge.js [hunk 3]
```diff
@@ -25,51 +24,255 @@ export default function BadgePage({ generatedDocGen }: {| generatedDocGen: DocGe
...
         `}
           />
         </MainSection.Subsection>
       </MainSection>
+      <MainSection name="Best practices">
+        <MainSection.Subsection columns={2}>
+          <MainSection.Card
+            cardSize="md"
+            type="do"
+            description="Align the badge to the top of large accompanying text. Center align for standard size text."
```
Comment (AlbertCarreras) [inline docs/pages/badge.js:48]: Instead  of _standard size text_ , could we specify sizes?

Comment (AlbertCarreras) [reply docs/pages/badge.js:48]: Maybe, I'd add a large Text  and a small text example in the same example. To exemplify both.

**Review by AlbertCarreras — APPROVED** [review]
LGTM, few nit comments.

File: packages/gestalt/src/Badge.js [hunk 4]
```diff
@@ -16,20 +15,33 @@ type Props = {|
...
+   */
+  type?: 'info' | 'error' | 'warning' | 'success' | 'neutral' | 'darkWash' | 'lightWash',
 |};
 
 /**
- * [Badge](https://gestalt.pinterest.systems/badge) is used to label text.
+ * [Badge](https://gestalt.pinterest.systems/badge) is a label that indicates status or importance. Badges should provide quick recognition.
  *
  * ![Badge light mode](https://raw.githubusercontent.com/pinterest/gestalt/master/cypress/integration/visual-test/__image_snapshots__/Badge%20%230.png)
```
Comment (AlbertCarreras) [inline packages/gestalt/src/Badge.js:27]: Here we could update the screenshots. A raw with all new types.

File: docs/pages/activationcard.js [hunk 5]
```diff
@@ -23,15 +23,15 @@ export default function ActivationCardPage({
           <MainSection.Card
             cardSize="md"
             type="do"
-            title="When to Use"
+            title="When to use"
```
Comment (dangerismycat) [inline docs/pages/activationcard.js:26]: Can these unrelated changes be pulled out to a separate PR to make it easier to review the relevant changes?

File: docs/pages/badge.js [hunk 6]
```diff
@@ -25,51 +24,255 @@ export default function BadgePage({ generatedDocGen }: {| generatedDocGen: DocGe
...
+              </Box>
+            </Mask>
+`}
+          />
+        </MainSection.Subsection>
+        <MainSection.Subsection columns={2}>
+          <MainSection.Card
+            cardSize="md"
+            type="do"
+            description="Use established color patterns so users can quickly scan an identify sentiment. However, the badge text should always clearly indicate the sentiment, so color is not the sole indicator of information."
```
Comment (dangerismycat) [inline docs/pages/badge.js:73]: typo: "…quickly scan _and_ identify…"

File: docs/pages/badge.js [hunk 7]
```diff
@@ -25,51 +24,255 @@ export default function BadgePage({ generatedDocGen }: {| generatedDocGen: DocGe
...
+            defaultCode={`
+            <Table accessibilityLabel="Badge color patterns">
```
Comment (dangerismycat) [inline docs/pages/badge.js:75]: Changing this to a function component will allow DRYing out the heavily-repeated hardcoded content

File: docs/pages/badge.js [hunk 8]
```diff
@@ -25,51 +24,255 @@ export default function BadgePage({ generatedDocGen }: {| generatedDocGen: DocGe
...
+                  </Table.Cell>
+                </Table.Row>
+              </Table.Body>
+            </Table>
+            `}
+          />
+          <MainSection.Card
+            cardSize="md"
+            type="don't"
+            description="Use complex or verbose language. Instead use a single, scannable word. For example, New."
```
Comment (dangerismycat) [inline docs/pages/badge.js:118]: Grammar nit (here and below): this uses a comma where it should use a colon (and probably quotes) in "For example: 'New'."

File: docs/pages/badge.js [hunk 9]
```diff
@@ -25,51 +24,255 @@ export default function BadgePage({ generatedDocGen }: {| generatedDocGen: DocGe
...
+          Badge is available in five styles. Each \`type\` represents a messaging sentiment.
 
-      <Example
-        description="
-    Larger text example rendered with a top positioned `Badge`."
-        name="Example: positioning"
-        defaultCode={`
-  <Text size="600">Ads & Campaigns <Badge text="Beta" position="top"/></Text>
+          1. **Info (default)**
+          Communicates helpful information or an important attribute. For example, new, help.
```
Comment (dangerismycat) [inline docs/pages/badge.js:143]: These badge examples should all be in sentence case

File: docs/pages/badge.js [hunk 10]
```diff
@@ -25,51 +24,255 @@ export default function BadgePage({ generatedDocGen }: {| generatedDocGen: DocGe
...
+          Indicates a constructive or successful state. For example, available, completed, approved, added.
+
+          3. **Warning**
+          Communicates cautionary or time-sensitive information to the user. For example, busy, missing, warning.
+
+          4. **Error**
+          Informs the user of problems or errors that require potential action to correct. For example, deleted, cancelled.
+
+          5. **Neutral**
+          Indicates a general, non-critical status update. For example, unavailable or not started.
```
[the remark this answers is not in the document]
Comment (dangerismycat) [reply docs/pages/badge.js:155]: +1

File: docs/pages/badge.js [hunk 11]
```diff
@@ -25,51 +24,255 @@ export default function BadgePage({ generatedDocGen }: {| generatedDocGen: DocGe
...
+ `}
+        >
+          <MainSection.Card
+            defaultCode={`
+  <Table accessibilityLabel="Type examples">
```
Comment (dangerismycat) [inline docs/pages/badge.js:160]: ditto here on using a function component to DRY out the markup

File: docs/pages/badge.js [hunk 12]
```diff
@@ -25,51 +24,255 @@ export default function BadgePage({ generatedDocGen }: {| generatedDocGen: DocGe
...
+          2. **Over media - Dark wash**
+          The dark wash badge should be used over media that is light or utilizes a light gradient overlay.
+`}
+        >
+          <MainSection.Card
+            defaultCode={`
+<Flex gap={4}>
+  <Mask height={150} width={280} rounding={2}>
+    <Image alt="Botanical art in coral and green" fit="cover" src="https://i.ibb.co/cbjgZft/img-door.jpg" naturalWidth={1} naturalHeight={1}/>
+    <Box position="absolute" top>
```
Comment (dangerismycat) [inline docs/pages/badge.js:234]: This isn't necessary; elements to be overlaid on Image should be children of the component, and absolute positioning is built-in

File: docs/pages/badge.js [hunk 13]
```diff
@@ -25,51 +24,255 @@ export default function BadgePage({ generatedDocGen }: {| generatedDocGen: DocGe
...
+<Flex gap={4} direction="column">
+  <Text size="300">Ads & Campaigns <Badge text="New" /></Text>
+  <Text size="600">Ads & Campaigns <Badge text="Beta" position="top"/></Text>
+</Flex>
+`}
+          />
+        </MainSection.Subsection>
+        <MainSection.Subsection
+          title="Within other components"
+          description="Components like [Module](/module) and [Dropdown](/dropdown) support Badges within the component."
```
Comment (dangerismycat) [inline docs/pages/badge.js:267]: This should be more explicit — something about how those components have Badge built-in, so using this component isn't necessary

File: docs/pages/badge.js [hunk 14]
```diff
@@ -78,8 +281,39 @@ function ModuleExample() {
...
+          />
+        </MainSection.Subsection>
+      </MainSection>
+      <MainSection name="Writing">
+        <MainSection.Subsection columns={2}>
+          <MainSection.Card
+            cardSize="md"
+            type="do"
+            description={`
+- Use a single word to describe the status of an element. For example, “new” not “new post.”
```
Comment (dangerismycat) [inline docs/pages/badge.js:294]: These examples should use sentence case

File: docs/pages/badge.js [hunk 15]
```diff
@@ -78,8 +281,39 @@ function ModuleExample() {
...
+- Where applicable, describe the status in past tense. For example, “archived” not “archive.”
 `}
-      />
+          />
+          <MainSection.Card
+            cardSize="md"
+            type="don't"
+            description={`
+- Use conflicting language with defined type sentiments. For example, the error badge should not say “complete.”
```
Comment (dangerismycat) [inline docs/pages/badge.js:302]: ditto here on sentence case

File: packages/gestalt-design-tokens/tokens/color/alias.json [hunk 16]
```diff
@@ -3,7 +3,7 @@
     "text": {
       "default": {
         "value": "{color.black.cosmicore.900.value}",
-        "darkValue": "{color.gray.roboflow.50.value}",
+        "darkValue": "{color.white.mochimalist.0.value}",
```
Comment (dangerismycat) [inline packages/gestalt-design-tokens/tokens/color/alias.json:6]: These changes also seem unrelated?

File: packages/gestalt/src/Badge.test.js [hunk 17]
```diff
@@ -8,9 +8,15 @@ it('Badge renders', () => {
...
   const { className } = instance.find((element) => element.type === 'div').props;
 
-  expect(className).toContain('blueBg');
+  expect(className).toContain('successBase');
+});
+
+it('should render with a wash', () => {
+  const component = create(<Badge text="Badge" type="lightWash" />);
+  const tree = component.toJSON();
+  expect(tree).toMatchSnapshot();
```
Comment (dangerismycat) [inline packages/gestalt/src/Badge.test.js:21]: Instead of a snapshot, can this verify the class as with the previous test?

File: packages/gestalt/src/Badge.test.js [hunk 18]
```diff
@@ -8,9 +8,15 @@ it('Badge renders', () => {
   expect(tree).toMatchSnapshot();
 });
 
-it('should render with white text and blue background', () => {
-  const instance = create(<Badge text="Badge" />).root;
+it('should render with proper styling', () => {
```
Comment (dangerismycat) [inline packages/gestalt/src/Badge.test.js:11]: Can we add these tests for all supported colors? (You can DRY that out by defining an array of objects with the `type` and expected className, then iterate over that array calling the test on each item)

File: docs/pages/activationcard.js (hunk 5 above)
Comment (ayeshakmaz) [reply docs/pages/activationcard.js:26]: I'm going to leave these for now since this PR is already approved, but will do for next time

File: packages/gestalt/src/Badge.test.js (hunk 18 above)
Comment (ayeshakmaz) [reply packages/gestalt/src/Badge.test.js:11]: This seems unnecessary if we have Visual regression tests, no?

File: packages/gestalt-design-tokens/tokens/color/alias.json (hunk 16 above)
Comment (ayeshakmaz) [reply packages/gestalt-design-tokens/tokens/color/alias.json:6]: They're related to the Badge colors, as roboflow 50 is not accessible

File: docs/pages/badge.js (hunk 11 above)
Comment (ayeshakmaz) [reply docs/pages/badge.js:163]: I think leaving these examples simpler makes it easier for people to digest the code example, where we focus on the content rather than creating functions for these mini examples. I think it's easier to copy, paste and use the code when you don't need to worry about removing functions and state variables

Comment (ayeshakmaz) [reply docs/pages/badge.js:163]: It also makes it easier for our designers to make updates to the docs
